### DISCUSSION TITLE
Persistence before action execution

### DISCUSSION THREAD

**Discussion by ldadams** [statelyai/xstate#4293] [Q&A]
Hello xstate community! I have two kinda related questions.

1. I have a need to persist machines and rehydrate their state.  My statecharts have actions that emit events out to a stream to invoke services.  Services respond via events back to the stream which are picked up and sent back to the machine that emitted them.  The general flow is:
Inbound Event -> Get Machine -> Hydrate Machine -> Send event to Machine -> Actions emit events -> Persist State

The problem I am encountering is around side effects of actions.  Basically, services are responding to events before the previous state is persisted.  This causes a race condition where a machine is hydrated with previous state.

I am looking for a way to capture / queue all the actions of a transition so that I can persist the state before executing them.

2. I have seen references to using Event Sourcing to rehydrate machines.  Is there a way to rehydrate a machine without executing actions?  My actions are "expensive" and ideally they don't execute while rehydrating a machine event by event.

Thank you!!

**Comment by doender** [comment]
@ldadams I've got a similar use case, did ever find a good solution for this?

**Comment by davidkpiano** [comment]
You can use a pattern where you defer those actions:
```ts
// global actions array
const actionFns = [];

function deferActionFn(actionFn) {
  return () => actionFns.push(actionFn)
}

function executeActions() {
  actionFns.forEach(actionFn => actionFn())

const machine = setup({
  actions: {
    doSomething: deferActionFn(() => {
      // do something
    }),
    doSomethingElse: deferActionFn(() => {
      // ...
    })
  }
});

const actor = createActor(machine);

actor.subscribe(snapshot => {
  persistState(snapshot).then(executeActions);
});

actor.start();
```
